src/Client.php

```diff
@@ -344,6 +345,9 @@ private function applyOptions(RequestInterface $request, array &$options): Reque
         ];
 
         if (isset($options['headers'])) {
+            if (array_keys($options['headers']) === range(0, count($options['headers']) - 1)) {
+                throw new RequestException('The headers array must have header name as keys.', $request);
```

Comment:
My understanding is that this PR is not yet in any release: Should this exception be changed to `InvalidArgumentException` instead? I seems to be more fitting.

Reply:
I think this was chosen for consistency with what similar code here does. I'm open to review this for Guzzle 8.

Reply:
Looking further below within this method I am just seeing `InvalidArgumentException` being thrown. That was my primary motivation to suggest the change.

e.g.

https://github.com/guzzle/guzzle/blob/e38c668f3b565bf9d045f20460a9bdf415073bf8/src/Client.php#L437

and

https://github.com/guzzle/guzzle/blob/e38c668f3b565bf9d045f20460a9bdf415073bf8/src/Client.php#L356-L362